packages/asset-swapper/src/network/chain.ts

```diff
@@ -0,0 +1,314 @@
+                to: calls[0].opts.to,
+                value: calls[0].opts.value || ZERO_AMOUNT,
+                data: calls[0].opts.data,
+                overrides: calls[0].opts.overrides || {},
+            };
+        }
+        const callInfos = calls.map(c =>
+            ({
+                data: c.opts.data,
+                to: c.opts.to,
+                gas: new BigNumber(c.opts.gas || 0),
+                value: c.opts.value || ZERO_AMOUNT,
+            }),
+        );
+        return {
+            gas: calls.reduce((s, c) => (c.opts.gas || DEFAULT_CALL_GAS_LIMIT) + s, 0) || DEFAULT_CALL_GAS_LIMIT,
+            from: DEFAULT_CALLER_ADDRESS,
+            gasPrice: BigNumber.sum(...calls.map(c => c.opts.gasPrice || 0)),
+            to: DISPATCHER_CONTRACT.address,
+            value: BigNumber.sum(...calls.map(c => c.opts.value || 0)),
+            data: DISPATCHER_CONTRACT.dispatch(callInfos).getABIEncodedTransactionData(),
+            overrides: Object.assign(
+                { [DISPATCHER_CONTRACT.address]: { code: DISPATCHER_CONTRACT_BYTECODE } },
+                ...calls.map(c => c.opts.overrides),
+            ),
+        };
+    }
+}
+
+function *generateCallBatches(queue: QueuedEthCall[]): Generator<QueuedEthCall[]> {
+    let nextQueue = queue;
+    while (nextQueue.length > 0) {
+        let _queue = nextQueue;
+        nextQueue = [];
+        const batch = [];
+        for (const c of _queue) {
+            if (canBatchCallWith(c, batch)) {
+                batch.push(c);
+            } else {
+                nextQueue.push(c);
+            }
+        }
+        yield batch;
+    }
+}
+
+function resolveQueuedCall(c: QueuedEthCall, r: DispatchedCallResult): void {
+    if (r.success) {
+        c.accept(r.resultData);
+    } else {
+        c.reject(tryDecodeStringRevertErrorResult(r.resultData) || `EVM call reverted`);
+    }
+}
+
+function tryDecodeStringRevertErrorResult(rawResultData: Bytes): string | undefined {
+    if (rawResultData.startsWith('0x08c379a0')) {
+        const strLen = new BigNumber(hexUtils.slice(rawResultData, 4, 36)).toNumber();
+        return Buffer.from(
+            hexUtils.slice(rawResultData, 68, 68 + strLen).slice(2),
+            'hex',
+        ).toString();
+    }
+    return;
+}
+
+function canBatchCallWith(ethCall: QueuedEthCall, batch: QueuedEthCall[]): boolean {
+    const { overrides } = {
+        overrides: {},
+        ...ethCall.opts,
+    };
+    for (const c of batch) {
+        // Overrides must not conflict.
+        for (const addr of Object.keys(overrides)) {
+            const a = overrides[addr];
+            const b = (c.opts.overrides || {})[addr];
+            if (a && b && a.code !== b.code) {
+                return false;
+            }
+        }
+    }
+    return true;
+}
+
+function hashCall(c: ChainEthCallOpts): string {
+    return crypto.createHash('sha1').update(JSON.stringify(c)).digest('hex');
```

Comment:
Supposedly this is the fastest collision-resistant, out-of-the-box hash function.